cre.cpp

```diff
@@ -1392,6 +1392,48 @@ static int isXPointerInCurrentPage(lua_State *L) {
 	return 1;
 }
 
+static int hasXPointer(lua_State *L) {
+    CreDocument *doc = (CreDocument*) luaL_checkudata(L, 1, "credocument");
+    const char *xpointer_str = luaL_checkstring(L, 2);
+
+    bool found = true;
+    ldomXPointer xp = doc->dom_doc->createXPointer(lString16(xpointer_str));
+    lvPoint pt = xp.toPoint();
+    if (pt.y < 0) {
+        found = false;
+    }
+    lua_pushboolean(L, found);
+    return 1;
+}
+
+static int getImageDataFromPosition(lua_State *L) {
+    CreDocument *doc = (CreDocument*) luaL_checkudata(L, 1, "credocument");
+    int x = luaL_checkint(L, 2);
+    int y = luaL_checkint(L, 3);
+    lvPoint pt(x, y);
+    ldomXPointer ptr = doc->text_view->getNodeByPoint(pt);
+    if (ptr.isNull())
+        return 0;
+    LVStreamRef stream = ptr.getNode()->getObjectImageStream();
+    if (!stream.isNull()) {
+        unsigned size = stream->GetSize();
+        lvsize_t read_size = 0;
+        void *buffer = (void *)malloc(size);
```

Comment:
> > Well, my first attempt code was simpler:

> It's simpler to on the C side (callee), but more complicate/hacky on the Lua side (caller) ;)

I meant the sample code in my comment above, using `lua_newuserdata()` but _without your suggesting metatable stuff_ (that I've seen in doc, but that I dont really understand), **does work**, and appears to not need anything from the lua side, and in my tests was freed immediately (or may be that was just cause i was repeating it 1000 times).
So, if you were to confirm this sample code look fine to you, I can replace it, there and in `getCoverPageImageData()` , and remove both `ffi.C.free(data)` from credocument.lua.

But if indeed, there's the chance that it is not gc'ed when it goes out of scope, we might as well, for these huge amount of image data, keep the explicite C.free().